Metadata schema info

aerfrei opened the pull request:
changefeedccl: add schema info to enriched envelope source
This change adds the fields table_name, schema_name, database_name
and primary_keys to the source for changefeed messages with the
enriched envelope type. They are fetched for each target table
on starting the changefeed agregator and stored by table id.

Epic: [CRDB-8665](https://cockroachlabs.atlassian.net/browse/CRDB-8665)
Fixes: https://github.com/cockroachdb/cockroach/issues/139692
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/143422)
<!-- Reviewable:end -->

aerfrei commented:
Extended CI failure seems related. Seems like leaktest.AfterTest(t)() is finding something. Probably a leaking goroutine. Not sure how this would be introduced only in TestChangefeedEnriched but not TestChangefeedEnrichedSourceWithData. Trying to investigate. Any ideas?

asg0451 commented on pkg/ccl/changefeedccl/enriched_source_provider.go:
```diff
@@ -134,6 +189,24 @@ func (p *enrichedSourceProvider) KafkaConnectJSONSchema() kcjsonschema.Schema {
+	tableID := metadata.TableID
+	tableInfo, ok := p.sourceData.tableSchemaInfo[tableID]
+	if !ok {
+		return nil, fmt.Errorf("table %d not found in tableSchemaInfo", tableID)
```
nit: can you make this an AssertionFailedf

asg0451 commented on pkg/ccl/changefeedccl/enriched_source_provider.go:
```diff
@@ -134,6 +189,24 @@ func (p *enrichedSourceProvider) KafkaConnectJSONSchema() kcjsonschema.Schema {
+	p.jsonNonFixedData[fieldNameSchemaName] = json.FromString(tableInfo.schemaName)
+	p.jsonNonFixedData[fieldNameTableName] = json.FromString(tableInfo.tableName)
+
+	primaryKeysBuilder := json.NewArrayBuilder(len(tableInfo.primaryKeys))
```
nit: might be worth storing this info as json ahead of time. ditto the above, but `json.FromString` is basically a no-op so i dont think that's necessary

asg0451 commented on pkg/ccl/changefeedccl/enriched_source_provider.go:
```diff
@@ -144,6 +217,12 @@ func (p *enrichedSourceProvider) GetJSON(updated cdcevent.Row) (json.JSON, error
+	tableID := row.EventDescriptor.TableDescriptor().GetID()
+	tableInfo, ok := p.sourceData.tableSchemaInfo[tableID]
+	if !ok {
+		return nil, fmt.Errorf("table %d not found in tableSchemaInfo", tableID)
```
ditto

asg0451 commented on pkg/ccl/changefeedccl/enriched_source_provider.go:
```diff
@@ -290,6 +381,52 @@ var allFieldInfo = map[string]fieldInfo{
+		avroSchemaField: avro.SchemaField{
+			// TODO:(aerin) how do I use this right?
+			Name:       fieldNamePrimaryKeys,
+			SchemaType: []avro.SchemaType{avro.SchemaTypeNull, avro.SchemaTypeString},
```
try:
```
[]avro.SchemaType{avro.SchemaTypeNull, avro.ArrayType{SchemaType: avro.SchemaTypeArray, Items: avro.SchemaTypeString}},
```

asg0451 commented on pkg/ccl/changefeedccl/enriched_source_provider.go:
```diff
@@ -157,6 +236,14 @@ func (p *enrichedSourceProvider) GetAvro(
+		// TODO:(aerin) I'd like this to be an array and not a string, byt this causes an error on
+		// the call to newFunctionalRecord saying it's missing Items. I think I'm not using the schema
+		// correctly. Trying to figure out the right way to specify this.
+		dest[fieldNamePrimaryKeys] = goavro.Union(avro.SchemaTypeString, fmt.Sprintf(`[ "%s" ]`, strings.Join(tableInfo.primaryKeys, `", "`)))
```
try: 

```
goavro.Union(avro.SchemaTypeArray, tableInfo.primaryKeys)
```

asg0451 commented on pkg/ccl/changefeedccl/fetch_table_bytes.go:
```diff
@@ -107,6 +107,42 @@ func getTableDesc(
 	return desc, nil
 }
 
+func getDBDesc(
```
this file is for a pretty specific feature, maybe put these helpers somewhere else despite the naming similarity

aerfrei replied:
I put them in this file (except for getTableDesc which I'm still using from here). Seems like other functions that fetch the table descriptors aren't in helper files either. I figure if we build up enough helpers we can put them in a file and until then it's alright to have these two here.

asg0451 commented on pkg/ccl/changefeedccl/changefeed_test.go:
```diff
@@ -4433,6 +4442,89 @@ func TestChangefeedEnrichedSourceWithData(t *testing.T) {
+				assertPayloadsEnriched(t, foo, testCase.expectedRowsAfterSchemaChange, sourceAssertionAfterSchemaChange)
+			}
+
+			cdcTest(t, testFn, feedTestForceSink("kafka"))
```
can you test that it does the right thing for primary key changes too?

aerfrei committed:
changefeedccl: add schema info to enriched envelope source

This change adds the fields table_name, schema_name, database_name
and primary_keys to the source for changefeed messages with the
enriched envelope type. They are fetched for each target table
on starting the changefeed agregator and stored by table id.

Epic: CRDB-8665
Fixes: #139692
Release note: None

aerfrei commented:
After identifying the extended CI failure as stemming from the pubsub test feed, we filed this issue: https://github.com/cockroachdb/cockroach/issues/144102. Failure seems test only so we can merge anyway. PR should be ready to merge.

aerfrei commented:
Note that we're keeping in the changes to (p *pubsubFeed).Close() since they seem like improvements. 
I also left in splitting the huge test that verifies the source data (TestChangefeedEnrichedSourceWithData) since it will make it clearer if that fails on non-pubsub sinks.

asg0451 commented on pkg/ccl/changefeedccl/enriched_source_provider.go:
```diff
@@ -42,11 +53,47 @@ type enrichedSourceProvider struct {
+		return nil
+	})
+	if err != nil {
+		return make(map[descpb.ID]tableSchemaInfo), err
```
why not return nil, err?

asg0451 commented on pkg/ccl/changefeedccl/enriched_source_provider.go:
```diff
@@ -371,3 +500,34 @@ func init() {
+	var dbDescriptor catalog.DatabaseDescriptor
+	var schemaDescriptor catalog.SchemaDescriptor
+	f := func(ctx context.Context, txn descs.Txn) error {
+		byIDGetter := txn.Descriptors().ByIDWithoutLeased(txn.KV()).WithoutNonPublic().Get()
```
what does WithoutNonPublic mean here? are there some tables this would fail to see?

aerfrei replied:
The comment said "WithoutNonPublic configures the ByIDGetterBuilder to error on descriptors which are not in a public state: dropped, offline, etc." I thought it made sense that we would not want to see dropped or offline tables. If we try to start a changefeed against a dropped table, I'd think we wouldn't even successfully start the changefeed anyway. What do you think?

asg0451 commented on pkg/ccl/changefeedccl/enriched_source_provider.go:
```diff
@@ -371,3 +500,34 @@ func init() {
+		if err != nil {
+			return err
+		}
+		tableDescriptor = tableDesc
```
nit: predeclare err so you can set these three directly without an extra var

asg0451 commented on pkg/ccl/changefeedccl/enriched_source_provider.go:
```diff
@@ -371,3 +500,34 @@ func init() {
+		schemaDescriptor = schemaDesc
+		return nil
+	}
+	if err := execCfg.InternalDB.DescsTxn(ctx, f, isql.WithPriority(admissionpb.LowPri)); err != nil {
```
is this low priority? since this is required for the feed to start up and only run once, maybe it should be normal priority

asg0451 commented on pkg/ccl/changefeedccl/event_processing.go:
```diff
@@ -104,7 +105,14 @@ func newEventConsumer(
 		sourceData := enrichedSourceData{}
 		if encodingOpts.Envelope == changefeedbase.OptEnvelopeEnriched {
-			sourceData, err = newEnrichedSourceData(ctx, cfg, spec, sink.getConcreteType())
+			schemaInfo := make(map[descpb.ID]tableSchemaInfo)
```
not: prefer var declaration here, unless empty vs nil is meaningful to your code, which it shouldn't be ideally

asg0451 commented on pkg/ccl/changefeedccl/helpers_test.go:
```diff
@@ -1577,14 +1579,26 @@ func ChangefeedJobPermissionsTestSetup(t *testing.T, s TestServer) {
 	return enrichedSourceData{
 		jobID:              "test_id",
 		dbVersion:          "test_db_version",
-		clusterName:        "test_cluster_name",
```
why remove these?

aerfrei commented:
bors r=asg0451
